### PR TITLE
[Torch] Serialize and load NNCF transformations

### PR DESCRIPTION
**Pull request by daniil-lyakhov**
On top of #2584 #2595 
### Changes

* `serialize_transformations` and `load_transformations` functions are introduced:  `serialize_transformations` could serialize `PTTransformationLayout`  to a dict which could be serialized by json; serialized transformation could be recovered by the `load_transformations` function.
* `StatefullTorchModuleInterface` is introduced to make it possible to serialize all compression modules for quantization, sparisification, weights compression and  pruning algorithms
* Quantizers are created with scale shape specified in the quantization spec

### Reason for changes

* To make it possible to serialize/deserialize PT nncf transformations
* To make it possible to serialize/deserialize compression modules which are the part of each transformation
* To align scales shapes after `nncf.quantize` and quantizes initialization

### Related tickets

129586

### Tests

* tests/torch/test_serialization.py
* tests/torch/test_serialization.py
* tests/torch/test_serialization.py

### CODE REVIEW

File: nncf/torch/graph/transformations/serialization.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    for serialized_command in transformations_state[COMPRESSION_STATE_ATTR]:
+        command = load_command(serialized_command)
+        transformation_layout.register(command)
+
+    return transformation_layout
+
+
+def serialize_command(command: PTTransformationCommand) -> Dict[str, Any]:
+    if not isinstance(command, (PTSharedFnInsertionCommand, PTInsertionCommand)):
+        return {}
```
Comment (daniil-lyakhov) [inline nncf/torch/graph/transformations/serialization.py:53]: TODO: test this if

File: nncf/torch/graph/transformations/serialization.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+COMPRESSION_STATE_ATTR = "compression_state"
+
+
+class CompressionKeys(Enum):
+    SHARED_INSERTION_COMMAND = "SHARED_INSERTION_COMMAND"
+    INSERTION_COMMAND = "INSERTION_COMMAND"
+
+
+def serialize_transformations(transformations_layout: TransformationLayout) -> Dict[str, Any]:
+    transformation_commands = []
```
Comment (daniil-lyakhov) [inline nncf/torch/graph/transformations/serialization.py:33]: TODO: docstring

File: nncf/torch/layer_utils.py [hunk 3]
```diff
@@ -19,6 +24,33 @@
...
+    The state for attributes with type torch.nn.Parameter
+    is recovered from the model `state_dict`, so there is no need to keep them in the module state.
+    Modules should avoid implementation of `__call__` method and use `forward` method instead,
+    as torch functions called inside the `__call__` method could not be unambiguously
+    separated from the wrapped parent nncf module functions calls, thus nncf is unable to
+    identify target point for that call during transformations recovery process.
+    """
+
+    @abstractmethod
+    def get_state(self) -> Dict[str, Any]:
```
Comment (alexsu52) [inline nncf/torch/layer_utils.py:42]: What do you think about `get_config`, `from_config` to align with high-level API?

File: nncf/torch/graph/transformations/serialization.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    Serializes given transformation layout to a dict.
+
+    :param tranformation_layout: Given transformation layout.
+    :return: Serialized representation of given transformation layout as a dict.
+    """
+    transformation_commands = []
+    for command in transformations_layout.transformations:
+        serialized_command = serialize_command(command)
+        if serialized_command:
+            transformation_commands.append(serialized_command)
```
Comment (alexsu52) [inline nncf/torch/graph/transformations/serialization.py:39]:
```suggestion
        transformation_commands.append(serialized_command)
```

File: nncf/torch/layer_utils.py (hunk 3 above)
Comment (daniil-lyakhov) [reply nncf/torch/layer_utils.py]: Done

File: nncf/torch/layer_utils.py [hunk 5]
```diff
@@ -19,6 +24,33 @@
 COMPRESSION_MODULES = Registry("compression modules")
 
 
+class StatefullTorchModuleInterface(ABC):
```
Comment (alexsu52) [inline nncf/torch/layer_utils.py:27]: IMHO: This is a non-intuitive class name. Please think about more suitable name.

May be:
`IModuleConfig`, `ModuleConfigInterface`, `StatefullModuleInterface`...

**Review by alexsu52 — COMMENTED** [review]
LGTM

File: nncf/torch/layer_utils.py (hunk 5 above)
Comment (daniil-lyakhov) [reply nncf/torch/layer_utils.py]: I picked `StatefullModuleInterface`, renamed

**Review by alexsu52 — APPROVED** [review]

**Pull request merged**